libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -423,39 +432,61 @@ AP_GPS::update(void)
         if (state[i].status != NO_GPS) {
             num_instances = i+1;
         }
-        if (_auto_switch) {            
-            if (i == primary_instance) {
-                continue;
-            }
-            if (state[i].status > state[primary_instance].status) {
-                // we have a higher status lock, change GPS
-                primary_instance = i;
-                continue;
-            }
-
-            bool another_gps_has_1_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 1);
+        if (i == primary_instance) {
+            continue;
+        }
 
-            if (state[i].status == state[primary_instance].status && another_gps_has_1_or_more_sats) {
+        // work out which GPS is the primary
+        if (now - _last_instance_swap_ms < 20000 &&
+                state[primary_instance].status >= GPS_OK_FIX_3D) {
+            // don't allow switches too often except if primary is in trouble
+            continue;
+        }
 
-                uint32_t now = AP_HAL::millis();
-                bool another_gps_has_2_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 2);
+        switch (_auto_switch) {
+        default:
+        case GPS_AUTO_SWITCH_DISABLED:
+            break;
 
-                if ( (another_gps_has_1_or_more_sats && (now - _last_instance_swap_ms) >= 20000) ||
-                     (another_gps_has_2_or_more_sats && (now - _last_instance_swap_ms) >= 5000 ) ) {
-                // this GPS has more satellites than the
-                // current primary, switch primary. Once we switch we will
-                // then tend to stick to the new GPS as primary. We don't
-                // want to switch too often as it will look like a
-                // position shift to the controllers.
-                primary_instance = i;
-                _last_instance_swap_ms = now;
-                }
+        case GPS_AUTO_SWITCH_MORE_SATS:
+            if (state[i].status > state[primary_instance].status ||
+                (state[i].num_sats >= state[primary_instance].num_sats + 2)) {
+                // there is a higher status lock or 2+ sats
+                primary_instance_change_request = i;
             }
-        } else {
-            primary_instance = 0;
+            break;
+
+        case GPS_AUTO_SWITCH_VZ:
+            float pri = fabsf(state[primary_instance].velocity.z);
+            float sec = fabsf(state[i].velocity.z);
+            if (pri > _vz_switch_threshold && // above threshold
+                sec < _vz_switch_threshold && // below threshold
+                fabsf(pri - sec) > 0.1f && // diverged by >10%. A little bit of difference is normal so lets reject that
```

Comment:
Isn't this just a difference of 0.1 m/s? The whole check seems based around the premise that the GPS reporting less velocity.z is always more accurate. It seems to me that it would make sense to look at the GPS that is reporting a higher confidence in it's velocity's rather then the minimal one. (Failure of a GPS could be holding a smaller velocity then the aircraft is experiencing, particularly if someone got the filter wrong on the dynamics its experiencing). It should be easy to review some logs where you have been switching on this and see if that would work as a better switching principle